velox/dwio/text/writer/TextWriter.cpp

```diff
@@ -0,0 +1,132 @@
+    const TypePtr& type,
+    vector_size_t row) {
+  if (columnDecodedData->isNullAt(row)) {
+    return NULLDATA;
+  }
+
+  switch (type->kind()) {
+    case TypeKind::BOOLEAN:
+      return std::to_string(columnDecodedData->valueAt<bool>(row));
+    case TypeKind::TINYINT:
+      return std::to_string(columnDecodedData->valueAt<int8_t>(row));
+    case TypeKind::SMALLINT:
+      return std::to_string(columnDecodedData->valueAt<int16_t>(row));
+    case TypeKind::INTEGER:
+      return std::to_string(columnDecodedData->valueAt<int32_t>(row));
+    case TypeKind::BIGINT:
+      return std::to_string(columnDecodedData->valueAt<int64_t>(row));
+    case TypeKind::REAL:
+      return std::to_string(columnDecodedData->valueAt<float>(row));
+    case TypeKind::DOUBLE:
+      return std::to_string(columnDecodedData->valueAt<double>(row));
+    case TypeKind::VARCHAR:
+      return columnDecodedData->valueAt<StringView>(row).getString();
+    case TypeKind::TIMESTAMP:
+      return columnDecodedData->valueAt<Timestamp>(row).toString();
+    case TypeKind::VARBINARY:
+      return columnDecodedData->valueAt<StringView>(row).getString();
+    // TODO
+    case TypeKind::ARRAY:
+    case TypeKind::MAP:
+    case TypeKind::ROW:
+    case TypeKind::UNKNOWN:
+    default:
+      VELOX_NYI("{} is not supported yet in TextWriter", type->kind());
+  }
+}
+
+TextWriter::TextWriter(
+    std::unique_ptr<dwio::common::FileSink> sink,
+    const WriterOptions& options,
+    std::shared_ptr<memory::MemoryPool> pool,
+    const RowTypePtr schema)
+    : pool_(pool), schema_(schema) {
+  bufferedWriter_ = std::make_unique<BufferedWriter>(std::move(sink), pool);
+}
+
+/**
+ *
+ */
+void TextWriter::write(const VectorPtr& data) {
+  VELOX_CHECK(
+      data->encoding() == VectorEncoding::Simple::ROW,
+      "Text writer expects row vector input");
+  VELOX_USER_CHECK(
+      data->type()->equivalent(*schema_),
+      "The file schema type should be equal with the input row vector type.");
+  const RowVector* dataRowVector = data->as<RowVector>();
+
+  const auto columnSize = dataRowVector->childrenSize();
+
+  std::vector<std::shared_ptr<DecodedVector>> decodedData;
+  for (size_t column = 0; column < columnSize; ++column) {
+    auto decodedColumn = std::make_shared<DecodedVector>(DecodedVector(
+        *dataRowVector->childAt(column),
+        SelectivityVector(dataRowVector->size())));
+    decodedData.push_back(decodedColumn);
+  }
+
+  for (vector_size_t row = 0; row < data->size(); ++row) {
+    for (size_t column = 0; column < columnSize; ++column) {
+      if (column != 0) {
+        bufferedWriter_->write(SOH);
+      }
+      auto columnData =
+          getColumnData(decodedData.at(column), schema_->childAt(column), row);
+      bufferedWriter_->write(columnData.c_str(), columnData.length());
+    }
+    bufferedWriter_->write(NEWLINE);
+  }
+}
+
+void TextWriter::flush() {
+  // Do nothing.
+}
+
+void TextWriter::close() {
+  bufferedWriter_->close();
```

Comment:
It is a bit dangerous to do flush in buffered writer dtor so we add abort method to bufferedWriter_ and call in TextWriter::abort(). Given that, we could just do buffer size zero check in bufferedWriter_ dtor? Thanks!